jax/source_info_util.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import inspect
+import os.path
+import threading
+from typing import List, NamedTuple, Optional
+
+class Frame(NamedTuple):
+  file_name: str
+  line_num: int
+  function_name: str
+
+SourceInfo = List[Frame]
+
+
+def _get_stacktrace() -> List[Frame]:
+  return [Frame(f.filename, f.lineno, f.function) for f in inspect.stack(0)]
+
+
+_jax_path = os.path.dirname(_get_stacktrace()[0].file_name)
+
+def user_frame(source_info: Optional[SourceInfo]) -> Optional[Frame]:
+  """Heuristic that guesses the identity of the user's code in a stack trace."""
+  try:
+    # Guess that the user's frame is the innermost stack frame that isn't in the
+    # jax source tree.
+    return next(x for x in (source_info or [])
+                if not x.file_name.startswith(_jax_path))
+  except StopIteration:
+    return None
```

Comment:
Instead of the `try`/`except` I think we can write

```python
return next((x for x in (source_info or []) if not x.file_name.startswith(_jax_path)), None)
```

Reply:
Cool, I didn't know `next` supported a default argument.